### PR TITLE
Managed Group Access Instructions

### PR DESCRIPTION
**Pull request by marctalbott**
Ticket: [GAWB-3764](https://broadinstitute.atlassian.net/browse/GAWB-3764) 
- move access instructions to Sam

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

File: src/main/resources/swagger/api-docs.yaml [hunk 1]
```diff
@@ -157,6 +157,30 @@ paths:
       tags:
         - Admin
 
+  /api/admin/groups/v1/{groupName}/accessInstructions:
```
Comment (dvoet) [inline src/main/resources/swagger/api-docs.yaml]: I know we are reproducing existing functionality... do we want to keep this a system admin feature only? Should we allow group admins to control this on their own?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala [hunk 2]
```diff
@@ -17,13 +17,23 @@ import scala.util.{Failure, Success, Try}
...
+        new Attribute(Attr.groupUpdatedTimestamp, formattedDate(new Date())),
+        new Attribute(Attr.accessInstructions, accessInstructions)
+      ) ++ membersAttribute
+
+      case None => Seq(
+        new Attribute("objectclass", "top", "workbenchGroup"),
+        new Attribute(Attr.email, group.email.value),
+        new Attribute(Attr.groupUpdatedTimestamp, formattedDate(new Date()))
+      ) ++ membersAttribute
+    }
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala]: too much code duplication above

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutes.scala [hunk 3]
```diff
@@ -33,28 +33,38 @@ trait ManagedGroupRoutes extends UserInfoDirectives with SecurityDirectives with
...
+            handleDeleteGroup(managedGroup, userInfo)
+          }
         } ~
         pathPrefix("requestAccess") {
           post {
             handleRequestAccess(managedGroup, userInfo)
           }
         } ~
+        path("accessInstructions") {
+          post {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutes.scala]: I think this should be a put

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/model/SamModel.scala [hunk 4]
```diff
@@ -104,5 +107,6 @@ case class AccessPolicyResponseEntry(policyName: AccessPolicyName, policy: Acces
 case class BasicWorkbenchGroup(id: WorkbenchGroupName, members: Set[WorkbenchSubject], email: WorkbenchEmail) extends WorkbenchGroup
 
 case class ManagedGroupMembershipEntry(groupName: ResourceId, role: AccessPolicyName, groupEmail: WorkbenchEmail)
+case class ManagedGroupAccessInstructions(groupName: String, instructions: String)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/model/SamModel.scala]: why is groupName in here? Can this be
`case class ManagedGroupAccessInstructions(value: String) extend ValueObject`

**Review by dvoet — CHANGES_REQUESTED** [review]
having the group name in the ManagedGroupAccessInstructions object is problematic because it is already specified in the URL path... which is right?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala [hunk 5]
```diff
@@ -128,15 +128,27 @@ class ManagedGroupService(private val resourceService: ResourceService, private
...
-    val resourceAndPolicyName = ResourceAndPolicyName(Resource(ManagedGroupService.managedGroupTypeName, resourceId), ManagedGroupService.adminPolicyName)
-    accessPolicyDAO.listFlattenedPolicyMembers(resourceAndPolicyName).map { users =>
-      val notifications = users.map { recipientUserId =>
-        Notifications.GroupAccessRequestNotification(recipientUserId, WorkbenchGroupName(resourceId.value).value, users, requesterUserId)
-      }
-
-      cloudExtensions.fireAndForgetNotifications(notifications)
+    getAccessInstructions(resourceId).map {
+      case accessInstructions: Some[String] =>
+        throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"Please follow special access instructions: $accessInstructions"))
```
Comment (Qi77Qi) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala]: can you do `Future.failed(new WorkbenchExceptionxxx)`? there was a similar discussion [here](https://github.com/broadinstitute/sam/pull/200#discussion_r215724582)

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala [hunk 6]
```diff
@@ -128,15 +128,27 @@ class ManagedGroupService(private val resourceService: ResourceService, private
...
   def requestAccess(resourceId: ResourceId, requesterUserId: WorkbenchUserId): Future[Unit] = {
-    val resourceAndPolicyName = ResourceAndPolicyName(Resource(ManagedGroupService.managedGroupTypeName, resourceId), ManagedGroupService.adminPolicyName)
-    accessPolicyDAO.listFlattenedPolicyMembers(resourceAndPolicyName).map { users =>
-      val notifications = users.map { recipientUserId =>
-        Notifications.GroupAccessRequestNotification(recipientUserId, WorkbenchGroupName(resourceId.value).value, users, requesterUserId)
-      }
-
-      cloudExtensions.fireAndForgetNotifications(notifications)
+    getAccessInstructions(resourceId).map {
+      case accessInstructions: Some[String] =>
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala]: `case Some(accessInstructions) =>`

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala [hunk 7]
```diff
@@ -329,4 +335,14 @@ class LdapDirectoryDAO(protected val ldapConnectionPool: LDAPConnectionPool, pro
     ldapConnectionPool.modify(petDn(petServiceAccount.id), modifications.asJava)
     petServiceAccount
   }
+
+  override def getManagedGroupAccessInstructions(groupName: WorkbenchGroupName): Future[Option[String]] = Future {
+    Option(ldapConnectionPool.getEntry(groupDn(groupName), Attr.accessInstructions)).map { entry =>
+      Option(entry.getAttributeValue(Attr.accessInstructions))
+    }.getOrElse(throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"$groupName not found")))
+  }
+
```
Comment (Qi77Qi) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala]:
```
for{
  entry <- Future{ldapConnectionPool.getEntry(groupDn(groupName), Attr.accessInstructions)}
  res <- Option(entry).fold(Future.failed(no group name found exception))(e => Option(e.getAttributeValue(Attr.accessInstructions)).fold(Future.failed(no attribute instruction exception))(x => Future.successful(x)))
} yield res
```
I'll do something like this. @dvoet might not like `fold`, but the main point is we'll know better where failure happens.

**Review by dvoet — APPROVED** [review]
1 minor nit

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala (hunk 7 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala]: I agree that I was confused by this block... but I am not less confused by this

I would try something like
```
    for {
      entry <- Future { ldapConnectionPool.getEntry(groupDn(groupName), Attr.accessInstructions) }
      res <- Option(entry).map {
        case None => Future.failed("no group name found exception")
        case Some(e) => Future.successful(Option(e.getAttributeValue(Attr.accessInstructions)))
    } yield res
```

**Review by Qi77Qi — APPROVED** [review]

**Pull request merged**